xinference/model/llm/ggml/ctransformers.py

```diff
@@ -0,0 +1,282 @@
+        ctransformers_generate_config.setdefault("stream", None)
+        ctransformers_generate_config.setdefault("reset", True)
+
+        return ctransformers_generate_config
+
+    def load(self):
+        try:
+            from ctransformers import AutoModelForCausalLM
+        except ImportError:
+            error_message = "Failed to import module 'ctransformers'"
+            if self._is_darwin_and_apple_silicon():
+                system = "Metal"
+            else:
+                system = "CUDA"
+
+            installation_guide = [
+                f"Please make sure 'ctransformers' is installed and {system} accelerator is provided.",
+                f"You can install it by checking out the repository for command for {system} platform:"
+                f"https://github.com/marella/ctransformers",
+            ]
+
+            raise ImportError(f"{error_message}\n\n{''.join(installation_guide)}")
+
+        model_path = os.path.join(
+            self.model_path,
+            self.model_spec.model_file_name_template.format(
+                quantization=self.quantization
+            ),
+        )
+
+        self._model_type = self._determine_model_type()
+        self._llm = AutoModelForCausalLM.from_pretrained(
+            model_path_or_repo_id=model_path,
+            model_type=self._model_type,
+            config=self._ctransformer_model_config,
+        )
+
+    @classmethod
+    def match(cls, llm_family: LLMFamilyV1, llm_spec: LLMSpecV1) -> bool:
+        if llm_spec.model_format != "ggmlv3":
+            return False
+        if "StarCoder" not in llm_family.model_name:
+            return False
+        if "generate" not in llm_family.model_ability:
+            return False
+        return True
+
+    def _determine_model_type(self):
+        if self._model_family.model_name not in model_type_for_ctransformer:
+            raise ValueError(
+                "The current model is not supported, check your model name. "
+            )
+        return model_type_for_ctransformer[self._model_family.model_name]
+
+    def generate(
+        self, prompt: str, generate_config_raw: CtransformersGenerateConfig
+    ) -> Union[Completion, Iterator[CompletionChunk]]:
+        def generator_wrapper(
+            _prompt: str,
+            _max_new_tokens: Union[int, None],
+            _generate_config: CtransformersGenerateConfig,
+        ) -> Iterator[CompletionChunk]:
+            assert self._model_uid is not None
+            for _completion_chunk, _ in generate_stream(
+                model=self._model_uid,
+                model_ref=self._llm,
+                prompt=_prompt,
+                max_new_tokens=_max_new_tokens,
+                **_generate_config,
+            ):
+                yield _completion_chunk
+
+        generate_config = self._sanitize_generate_config(generate_config_raw)
+        max_new_tokens: Union[int, None]
+        if "max_tokens" in generate_config:
```

Comment:
`max_new_tokens = generate_config.pop("max_tokens", None)`

Reply:
updated in the latest version